src/Common/FileSegment.cpp

```diff
@@ -593,10 +500,62 @@ void FileSegment::completeImpl(std::lock_guard<std::mutex> & cache_lock, std::lo
         }
     }
 
-    if (!downloader_id.empty() && (isDownloaderImpl(segment_lock) || is_last_holder))
+    switch (download_state)
     {
-        LOG_TEST(log, "Clearing downloader id: {}, current state: {}", downloader_id, stateToString(download_state));
-        downloader_id.clear();
+        case State::SKIP_CACHE:
+        {
+            if (is_last_holder)
+                cache->remove(key(), offset(), cache_lock, segment_lock);
+
+            return;
+        }
+        case State::DOWNLOADED:
+        {
+            assert(downloaded_size == range().size());
+            assert(is_downloaded);
+            break;
+        }
+        case State::DOWNLOADING:
```

Comment:
Yes, so in this case nothing is actually done as you can see, only one assertion and then we return. This case is here for the case when `completeWithoutState -> completeBasedOnCurrentState` is used - which is called from `FileSegmentHolder's` destructor which is called always in destructor of read buffers. We see that state is DOWNLOADING and `is_last_holder == false && is_downloader == false` and do nothing. For `completeWithState` it will throw exception that it is not possible to complete with DOWNLOADING state.